### PR TITLE
Improved C++ Language Service handling of templates and ServiceVersion constants

### PR DESCRIPTION
**Pull request by LarryOsterman**
Improved the overall quality of the APIView code handling templated types.

Issues #2973 #2974 #2976

### CODE REVIEW

**Review by vhvb1989 — APPROVED** [review]
<3

File: src/dotnet/APIView/APIViewWeb/Languages/CppLanguageService.cs [hunk 1]
```diff
@@ -315,30 +318,40 @@ string GenerateUniqueMethodId(CppAstNode methodNode)
                 return bldr.ToString();
             }
 
-            NavigationItem ProcessClassNode(CppAstNode node, string parentName)
+            NavigationItem ProcessClassNode(CppAstNode node, string parentName, List<CppAstNode> templateParams = null)
             {
                 NavigationItem navigationItem = null;
 
-                // Skip empty forward declarations
-                if (node.inner == null)
```
Comment (praveenkuttappan) [inline src/dotnet/APIView/APIViewWeb/Languages/CppLanguageService.cs:323]: I think there was a specific ask to disable forward declaration.

Comment (LarryOsterman) [reply src/dotnet/APIView/APIViewWeb/Languages/CppLanguageService.cs:323]: IIRC, the problem is that the ClassTemplateDecl had a CxxRecordDecl with no child nodes, which meant that template specializations without fields wouldn't display.

**Comment by ahsonkhan** [comment]
Looks like this also fixes https://github.com/Azure/azure-sdk-tools/issues/1390

**Pull request merged**